Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs

```diff
@@ -0,0 +1,252 @@
+
+            foreach (var activity in activities)
+            {
+                if (activity.Date >= minDate && activity.Date <= maxDate)
+                {
+                    checkedActivities.Add(activity);
+                }
+                else
+                {
+                    _logger?.LogWarning($"Inconsistent activity date. Activity={APIHelper.JsonSerialize(activity)}");
+                }
+            }
+
+            activities = checkedActivities;
+        }
+
+        public void CheckActivityPaths(ref IEnumerable<ActivityInfo> activities, IEnumerable<Member> members)
+        {
+            foreach (var activity in activities)
+            {
+                if (activity.Paths != null && activity.Paths.Any()) continue;
+
+                var member = members.FirstOrDefault(x =>
+                    x.MentorId == activity.UserId || x.RootMentorId == activity.UserId);
+
+                if (member == null) continue;
+
+                activity.Paths = new List<string>(member.Paths);
+                _logger?.LogWarning($"Activity paths are empty. Getting from member's paths. Activity={APIHelper.JsonSerialize(activity)}");
+            }
+        }
+
+        public void CheckActivityUser(ref IEnumerable<ActivityInfo> activities, IEnumerable<Member> members)
+        {
+            var checkedActivities = new List<ActivityInfo>();
+
+            foreach (var activity in activities)
+            {
+                if (activity.UserId != null)
+                {
+                    if (members.Any(x => x.MentorId == activity.UserId || x.RootMentorId == activity.UserId))
+                    {
+                        checkedActivities.Add(activity);
+                        continue;
+                    }
+
+                    _logger?.LogWarning($"User with UserId={activity.UserId} is not found. Activity={APIHelper.JsonSerialize(activity)}");
+                }
+
+                if (activity.UserEmail == null)
+                {
+                    _logger?.LogWarning($"UserId and email are empty. Activity={APIHelper.JsonSerialize(activity)}");
+                    continue;
+                }
+
+                var id = GetUserIdByEmail(activity.UserEmail, members);
+                if (id == null)
+                {
+                    _logger?.LogWarning($"User with email={activity.UserEmail} is not found. Activity={APIHelper.JsonSerialize(activity)}");
+                    continue;
+                }
+
+                activity.UserId = id;
+                checkedActivities.Add(activity);
+            }
+
+            activities = checkedActivities;
+        }
+
+        public IEnumerable<SpActivity> Ensure(IEnumerable<SpActivity> spActivities, IEnumerable<ActivityInfo> activities, IEnumerable<Member> members)
+        {
+            try
+            {
+                _logger?.LogInformation("Ensuring activities");
+
+                var dict = InitDictionary(spActivities);
```

Comment:
Actually, all code which works with SpActivity might be moved into `EnsureService`. Here you may retrieve all existing activities. Inside `SyncActivityService` you may just call `Save` method of `EnsureService`